src/logic/helpers/validateSource.js

```diff
@@ -0,0 +1,18 @@
+import retryingFailedImports from '../../util/retryingFailedImports';
+import {validatedSource} from '../../actions/errors';
+
+function importValidations() {
+  return retryingFailedImports(() =>
+    import(
+      /* webpackChunkName: "mainAsync" */
+      '../../validations'
+    ),
+  );
+}
+
+export default async ({language, source, projectAttributes}, dispatch) => {
+  const validations = await importValidations();
+  const validate = validations[language];
+  const validationErrors = await validate(source, projectAttributes);
+  await dispatch(validatedSource(language, validationErrors));
```

Comment:
How come we’re awaiting the `dispatch` here? `dispatch` should be a void function; I wouldn’t expect it to return a promise…